Tribler/community/tunnel/hidden_community.py

```diff
@@ -437,22 +476,30 @@ def check_linked_e2e(self, messages):
     def on_linked_e2e(self, messages):
         for message in messages:
             cache = self.request_cache.pop(u"link-request", message.payload.identifier)
-            
             download = self.find_download(cache.info_hash)
             if download:
                 download.add_peer((self.circuit_id_to_ip(cache.circuit.circuit_id), CIRCUIT_ID_PORT))
             else:
                 self._logger.error('On linked e2e: could not find download!')
 
-    def create_introduction_point(self, info_hash, amount=1):
+            for download in self.trsession.get_downloads():
+                if cache.info_hash == self.get_lookup_info_hash(download.get_def().get_infohash()):
+                    download.add_peer((self.circuit_id_to_ip(cache.circuit.circuit_id), CIRCUIT_ID_PORT))
+                    break
+
+    def create_introduction_point(self, info_hash, amount=1, required_exit=None):
         # Create a separate key per infohash
         if info_hash not in self.session_keys:
             self.session_keys[info_hash] = self.crypto.generate_key(u"curve25519")
 
         def callback(circuit):
-            # We got a circuit, now let's create a introduction point
+            # We got a circuit, now let's create an introduction point
             circuit_id = circuit.circuit_id
-            self.my_intro_points[circuit_id].append(info_hash)
+            self._logger.debug("We got an introduction circuit ending in %s:%s" %
+                               (circuit.required_exit[0], circuit.required_exit[1]))
+            self.my_intro_points[circuit_id].append((info_hash))
+            self.seeding_introduction[info_hash] = (circuit_id, circuit.required_exit)
+            self.infohashes_ip_waiting_for_circuit.pop(info_hash)
```

Comment:
So if we want 5 introduction points, we're fine if we only actually construct one